examples/multithreading_two_way.py

```diff
@@ -0,0 +1,113 @@
+import time
+
+from qtpy.QtWidgets import (
+    QGridLayout,
+    QLabel,
+    QProgressBar,
+    QPushButton,
+    QWidget,
+)
+
+import napari
+from napari._qt.threading import ProgressWorker, thread_worker
+
+
+@thread_worker(worker_class=ProgressWorker)
+def two_way_communication_with_args(start, end):
+    """Both sends and receives values to & from the main thread.
+
+    Accepts arguments, puts them on the worker object.
+    Receives values from main thread with ``incoming = yield``
+    Optionally returns a value at the end
+    """
+
+    # optionally, do setup here... this happens in the main thread
+    # if you yield a dict, they will be added as attributes to the worker
+    yield {
+        'start_val': start,
+        'end_val': end,
+        'some_other_val': 'hi',
+        '__len__': end,  # note, the '__len__' key is required for ProgressWorker
+    }
+
+    # do computationally intensive work here
+    i = start
+    while i <= end:
+        time.sleep(0.1)
+        # incoming receives values from the main thread
+        # while yielding sends values back to the main thread
+        incoming = yield i
+        i = incoming if incoming is not None else i + 1
+
+    # do optional teardown here
+    return "done"
+
+
+class Controller(QWidget):
+    def __init__(self):
+        super().__init__()
+
+        layout = QGridLayout()
+        self.setLayout(layout)
+        self.status = QLabel('Click "Start"', self)
+        self.play_btn = QPushButton("Start", self)
+        self.abort_btn = QPushButton("Abort!", self)
+        self.reset_btn = QPushButton("Reset", self)
```

Comment:
Would be nice if pressing reset could send progress bar down to zero right away, instead it doesn't reset until continue is pressed. also reset doesn't seem to work after the whole example has finished and the "done" appears, i.e. i can't seem to run the whole thing twice which might be nice

amazing stuff though!

Reply:
resetting to zero immediately is not technically possible with threading... I mean, we could just "put" the slider to zero, but until the thread "checks in" there's really no way to know where the thread is at.  (There's a concept in web design called "optimistic updates" in which you set the GUI to what you "hope" will happen, and after the asynchronous call returns, you fix the GUI if you were wrong ... but I think for this example, that would unnecessarily hide the fact that asynchronous work is definitely a bit different).

Reusing workers is possible... but adds a level of difficulty (which is why I disable the button on finished).  I will look into that in a future PR

Reply:
ah interesting, maybe then just disable the reset button when the example is done?

Reply:
done